buildSrc/src/main/kotlin/Java9Modularity.kt

```diff
@@ -73,39 +75,42 @@ object Java9Modularity {
         }
     }
 
-    private fun Project.registerCompileModuleTask(taskName: String, compileTask: KotlinCompile, sourceFile: File, targetFile: File) =
+    private fun Project.registerCompileModuleTask(taskName: String, compileTask: KotlinCompile, sourceFile: File, targetDirectory: Provider<out Directory>) =
         tasks.register(taskName, JavaCompile::class) {
-            // Also add the module-info.java source file to the Kotlin compile task;
-            // the Kotlin compiler will parse and check module dependencies,
-            // but it currently won't compile to a module-info.class file.
-            compileTask.source(sourceFile)
```

Comment:
If you remove this line, please provide an alternative way. It was here to perform validation of the sources in accordance to the `module-info` file, and now nothing performs this validation.

Reply:
Can you tell me what exactly that did / checked?
What kind of bad change did that capture, so what would make the build fail with it?

Reply:
I'm sorry we lack documentation about this. Actually, Kotlin compiler is able to read `module-info.java` file and verify it (the comment says 'the Kotlin compiler will parse and check module dependencies'). It verifies that all package names in `exports` are correct and all `requires` dependencies are present (not so sure about latter, @ALikhachev ?).
What happens here is we add `module-info.java` location to the inputs of `KotlinCompile` task so Kotlin compiler can perform this verification.

Reply:
Hm, I see.
But I'm not sure whether there is a good solution, besides having the normal compile task compile with Java 8 toolchain and having a second set of compile tasks with Java 11 toolchain and jdk-release 9 / jvm-target 9 that is hooked into the `check` task dependencies. :-/

If I just add `-Xjdk-release=1.8` as free compiler arg into the existing (pre-my-PR) setup, there come a looot of errors about classes being in modules that are not depended upon.

If with the new (with-my-PR)  setup I use JVM toolchain 17 (to reproduce the problematic bytecode eventually) and add jdk-release 1.8 / jvm-target 1.8, the correct bytecode is produced but still those errors are not verified as the module infos seem to be ignored.

If I additionally add the `module-info.java` to the `KotlinCompile` task again, I get the same wall of errors I get if I just add "jdk-release 1.8" to the existing setup.

If I remove the "jdk-release" and only keep the "jvm-target", the verifications work again, but the wrong byte-code is produced. *grml*

Reply:
Let me try to sum this up: the problem is, when `module-info.java` is in the `KotlinCompile` task inputs AND `jdk-release=1.8` specified to use correct Java API, then there are a lot of errors. What this errors are, exactly? I think this scenario is popular for multi-release JARs and should be somewhat supported without much setup, but I may be wrong. Judging by https://stackoverflow.com/a/55143699/5438558, to solve similar problem in Java, specifying `module-path` is enough.

Reply:
However, after giving it a second thought, I'm not so sure we need this verification at all. We have `JavaCompile` task for module-info anyway — it will tell us if some of dependencies are missing and/or exported incorrectly, won't it?

Reply:
No, it wont actually.

The `JavaCompile` task just compiles the `module-info.java` file.
It will complain if there is an error within, like requiring a module that is not present, or exporting a package that does not exist.
But the intended verification happens when compiling the Kotlin files JPMS aware.
So if you for example remove `exports kotlinx.serialization;` from `core`, `formats/json` compilation will fail with the verification for using classes in non-exported packages.
And if you for example remove `requires transitive kotlinx.serialization.core;` from `formats/json`, its compilation will fail with the verification for using classes without requiring the module.

This verification still has some bugs like https://youtrack.jetbrains.com/issue/KT-60583 and https://youtrack.jetbrains.com/issue/KT-60582 for example.

I've added the alternative means for the validation that @ALikhachev requested now.
It needed some work-arounds due to shortcomings in the Kotlin Gradle Plugin and in the `module-info` handling.
But all are reported and documented with code comments.
And from my tests hopefully the same verifications done before are now again done.
Thank you very much @Tapchicoma for your thorough help in getting this set up.